pkg/util/webhookpatch.go

```diff
@@ -64,3 +75,108 @@ func PatchMutatingWebhookConfig(client admissionregistrationv1beta1client.Mutati
 	}
 	return err
 }
+
+const delayedRetryTime = time.Second
+
+// Moved out of injector main. Changes:
+// - pass the existing k8s client
+// - use the K8S root instead of citadel root CA
+// - removed the watcher - the k8s CA is already mounted at startup, no more delay waiting for it
+func PatchCertLoop(injectionWebhookConfigName, webhookName, caBundlePath string, client kubernetes.Interface, stopCh <-chan struct{}) {
+	// K8S own CA
+	caCertPem, err := ioutil.ReadFile(caBundlePath)
+	if err != nil {
+		log.Errorf("Skipping webhook patch, missing CA path %v", caBundlePath)
+		return
+	}
+
+	var retry bool
+	if err = PatchMutatingWebhookConfig(client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations(),
+		injectionWebhookConfigName, webhookName, caCertPem); err != nil {
+		log.Warna("Error patching Webhook ", err)
+		retry = true
+	}
+
+	shouldPatch := make(chan struct{})
+
+	watchlist := cache.NewListWatchFromClient(
+		client.AdmissionregistrationV1beta1().RESTClient(),
+		"mutatingwebhookconfigurations",
+		"",
+		fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", injectionWebhookConfigName)))
+
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1beta1.MutatingWebhookConfiguration{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldConfig := oldObj.(*v1beta1.MutatingWebhookConfiguration)
+				newConfig := newObj.(*v1beta1.MutatingWebhookConfiguration)
+
+				if oldConfig.ResourceVersion != newConfig.ResourceVersion {
+					for i, w := range newConfig.Webhooks {
+						if w.Name == webhookName && !bytes.Equal(newConfig.Webhooks[i].ClientConfig.CABundle, caCertPem) {
+							log.Infof("Detected a change in CABundle, patching MutatingWebhookConfiguration again")
+							shouldPatch <- struct{}{}
+							break
+						}
+					}
+				}
+			},
+		},
+	)
+	go controller.Run(stopCh)
+
+	go func() {
+		var delayedRetryC <-chan time.Time
+		if retry {
+			delayedRetryC = time.After(delayedRetryTime)
+		}
+
+		for {
+			select {
+			case <-delayedRetryC:
+				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
+					delayedRetryC = time.After(delayedRetryTime)
+				} else {
+					log.Infof("Retried patch succeeded")
+					delayedRetryC = nil
+				}
+			case <-shouldPatch:
+				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
+					if delayedRetryC == nil {
+						delayedRetryC = time.After(delayedRetryTime)
+					}
+				} else {
+					delayedRetryC = nil
+				}
+			}
+		}
+	}()
+}
+
+func doPatch(cs kubernetes.Interface, webhookConfigName, webhookName string, caCertPem []byte) (retry bool) {
+	client := cs.AdmissionregistrationV1beta1().MutatingWebhookConfigurations()
+	if err := PatchMutatingWebhookConfig(client, webhookConfigName, webhookName, caCertPem); err != nil {
```

Comment:
We can unify the mutating/validating webhookconfiguration reconcile in a separate pr. Now one is patch and the other is update

Reply:
this is to reuse original code that validationwebhook is using `update` while mutationwebhook is using `patch`